src/ConstantInterval.cpp

```diff
@@ -0,0 +1,660 @@
+    return a + ConstantInterval::single_point(b);
+}
+
+ConstantInterval operator-(const ConstantInterval &a, int64_t b) {
+    return a - ConstantInterval::single_point(b);
+}
+
+ConstantInterval operator/(const ConstantInterval &a, int64_t b) {
+    return a / ConstantInterval::single_point(b);
+}
+
+ConstantInterval operator*(const ConstantInterval &a, int64_t b) {
+    return a * ConstantInterval::single_point(b);
+}
+
+ConstantInterval operator%(const ConstantInterval &a, int64_t b) {
+    return a % ConstantInterval::single_point(b);
+}
+
+ConstantInterval min(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result;
+    result.max_defined = a.max_defined || b.max_defined;
+    result.min_defined = a.min_defined && b.min_defined;
+    if (a.max_defined && b.max_defined) {
+        result.max = std::min(a.max, b.max);
+    } else if (a.max_defined) {
+        result.max = a.max;
+    } else if (b.max_defined) {
+        result.max = b.max;
+    }
+    if (a.min_defined && b.min_defined) {
+        result.min = std::min(a.min, b.min);
+    }
+    return result;
+}
+
+ConstantInterval min(const ConstantInterval &a, int64_t b) {
+    ConstantInterval result = a;
+    if (result.max_defined) {
+        result.max = std::min(a.max, b);
+    } else {
+        result.max = b;
+        result.max_defined = true;
+    }
+    if (result.min_defined) {
+        result.min = std::min(a.min, b);
+    }
+    return result;
+}
+
+ConstantInterval max(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result;
+    result.min_defined = a.min_defined || b.min_defined;
+    result.max_defined = a.max_defined && b.max_defined;
+    if (a.min_defined && b.min_defined) {
+        result.min = std::max(a.min, b.min);
+    } else if (a.min_defined) {
+        result.min = a.min;
+    } else if (b.min_defined) {
+        result.min = b.min;
+    }
+    if (a.max_defined && b.max_defined) {
+        result.max = std::max(a.max, b.max);
+    }
+    return result;
+}
+
+ConstantInterval max(const ConstantInterval &a, int64_t b) {
+    ConstantInterval result = a;
+    if (result.min_defined) {
+        result.min = std::max(a.min, b);
+    } else {
+        result.min = b;
+        result.min_defined = true;
+    }
+    if (result.max_defined) {
+        result.max = std::max(a.max, b);
+    }
+    return result;
+}
+
+ConstantInterval abs(const ConstantInterval &a) {
+    ConstantInterval result;
+    if (a.min_defined && a.max_defined && a.min != INT64_MIN) {
+        result.max_defined = true;
+        result.max = std::max(-a.min, a.max);
+    }
+    result.min_defined = true;
+    if (a.min_defined && a.min > 0) {
+        result.min = a.min;
+    } else {
+        result.min = 0;
+    }
+
+    return result;
+}
+
+ConstantInterval operator<<(const ConstantInterval &a, const ConstantInterval &b) {
+    // Try to map this to a multiplication and a division
```

Comment:
I'm not sure I follow this method entirely, more explanation in comments would be helpful

Reply:
It made sense to me at the time, but it doesn't now. I'll rewrite it.